### PR TITLE
Implement RANLUX++ generator

### PR DESCRIPTION
**Pull request by hahnjo**
RANLUX++ is an LCG equivalent of RANLUX using 576 bit numbers.

Described in
A. Sibidanov, *A revision of the subtract-with-borrow random numbergenerators*,
*Computer Physics Communications*, 221(2017), 299-303,
preprint https://arxiv.org/pdf/1705.03123.pdf

The code is loosely based on the Assembly implementation by A. Sibidanov
available at https://github.com/sibidanov/ranluxpp/.

### CODE REVIEW

**Comment by hahnjo** [comment]
My machine finally completed full testing with the TestU01 suite (via https://github.com/lmoneta/testrng):
```
========= Summary results of BigCrush =========

 Version:          TestU01 1.2.3
 Generator:        TRandomRanluxpp (p = 2048)
 Number of statistics:  160
 Total CPU time:   04:56:59.27

 All tests were passed
```

Performance-wise, I tested summing up 1e9 random doubles and each number takes 29ns on average. As expected, that is worse than the assembly implementation (approx. 9ns), but at least twice as fast compared to RANLUX in `TRandom1` with ca. 62ns.

File: math/mathcore/src/RanluxppEngineImpl.cxx
Comment (Axel-Naumann) [inline math/mathcore/src/RanluxppEngineImpl.cxx:184]: Maybe you prefer this? Up to you!

```suggestion
   union {
     double drandom;
     uint64_t irandom;
   } castUnion;
   castUnion.irandom = exp | bits;
   
   // Shift to the right interval of [0, 1).
   return castUnion.drandom - 1;
```

File: math/mathcore/src/mulmod.h
Comment (Axel-Naumann) [inline math/mathcore/src/mulmod.h:3]: Super impressive this file, congrats, @hahnjo !

File: math/mathcore/test/RanluxppEngine.cxx
Comment (Axel-Naumann) [inline math/mathcore/test/RanluxppEngine.cxx:24]: Awesome!

**Review by Axel-Naumann — COMMENTED** [review]
Up to @lmoneta but really: fantastic work, Jonas!

**Comment by lmoneta** [comment]
Great work Jonas! I will work at the detail and try to understand the reason for skip(11). 
Good you have run testU01. 29 ns/number is still an excellent time!

File: math/mathcore/src/RanluxppEngineImpl.cxx
Comment (hahnjo) [reply math/mathcore/src/RanluxppEngineImpl.cxx:134]: Yes, an oversight. I've changed all of the arithmetic to `int`, the code doesn't need the extra range provided by `unsigned`.

File: math/mathcore/src/RanluxppEngineImpl.cxx
Comment (hahnjo) [reply math/mathcore/src/RanluxppEngineImpl.cxx:175]: Partly yes: The way I implemented `RanluxppEngineImpl` allows to get custom amounts of random bits via the `w` template argument. For the exposed `RanluxppEngine`, we care mostly about `double`s which need 52 bits (well 53 bits according to the RANLUX++ paper, but this wastes more bits at the end of each block). In the future the flexibility of `RanluxppEngineImpl` could be used to create an engine that produces single precision `float`s with only 23/24 bits.

Comment (Axel-Naumann) [reply math/mathcore/src/RanluxppEngineImpl.cxx:175]: Argh of course! I should have seen that. Thanks for explaining, Jonas!

File: math/mathcore/src/RanluxppEngineImpl.cxx
Comment (Axel-Naumann) [reply math/mathcore/src/RanluxppEngineImpl.cxx:99]: https://wesmckinney.com/blog/avoid-unsigned-integers/ is a nice summary of the concerns.

Comment (hahnjo) [reply math/mathcore/src/RanluxppEngineImpl.cxx:99]: I've changed the code to `int`, added some `assert`s and a cast in `n < left` below to avoid a compiler warning. I'd like to keep the `unsigned` for the carry bits in `mulmod.h` to reduce the number of casts in conjunction with `uint64_t`.

File: math/mathcore/src/RanluxppEngineImpl.cxx
Comment (hahnjo) [reply math/mathcore/src/RanluxppEngineImpl.cxx:184]: I was hoping that I could just cast somehow, but haven't found a way. I went with the `union` but made it anonymous.

**Comment by hahnjo** [comment]
> 29 ns/number is still an excellent time!

I tried a bit and found a few tweaks to get the average time of the portable and standard-compliant version down to 24ns. Using the extension `__int128` (if supported), I can improve this to 13ns which is pretty close to the assembly version. I'm still experimenting a bit, will submit the optimizations as a follow-up PR once this one is merged.

File: math/mathcore/inc/Math/RanluxppEngine.h
Comment (vgvassilev) [inline math/mathcore/inc/Math/RanluxppEngine.h:27]: Does this class make sense to be derived from? If not maybe add final to hint the compiler to devirtualize the virtual methods.

**Comment by hahnjo** [comment]
@phsft-bot build just on windows10/cxx14

**Comment by hahnjo** [comment]
@phsft-bot build just on windows10/cxx14

**Review by lmoneta — APPROVED** [review]
Looks great! Please go ahead Jonas and merge it

**Pull request merged**
